Simplify `attribute_groups`

nnethercote opened the pull request:
It's more complicated than it needs to be.

r? @jdonszelmann

nnethercote committed:
Simplify `Accepts`.

There only needs to be one `Fn` per symbol, not multiple.

nnethercote committed:
Introduce some typedefs to improve readability.

nnethercote committed:
Fix up some comments.

Some are too long (> 100 chars), some are too short, some are missing
full stops, some are missing upper-case letters at the start of
sentences.

jieyouxu commented on compiler/rustc_attr_parsing/src/context.rs:
```diff
@@ -263,21 +267,19 @@ impl<'sess> AttributeParser<'sess> {
-                        // to this assertion.
+                        // If we're here, we must be compiling a tool attribute... Or someone
+                        // forgot to parse their fancy new attribute. Let's warn them in any case.
+                        // If you are that person, and you really your attribute should remain
+                        // unparsed, carefully read the documentation in this module and if you
```
Nit:

```suggestion
                        // If you are that person, and you really think your attribute should remain
                        // unparsed, carefully read the documentation in this module and if you
```

nnethercote replied:
:+1:

jdonszelmann commented on compiler/rustc_attr_parsing/src/context.rs:
```diff
@@ -26,24 +26,29 @@ macro_rules! attribute_groups {
 
                     for (k, v) in <$names>::ATTRIBUTES {
-                        accepts.entry(*k).or_default().push(Box::new(|cx, args| {
+                        let old = accepts.insert(*k, Box::new(|cx, args| {
```
fair, used to be a vec I think but this is better

jdonszelmann commented on compiler/rustc_attr_parsing/src/attributes/mod.rs:
```diff
@@ -33,7 +33,7 @@ pub(crate) mod transparency;
 
 type AcceptFn<T> = fn(&mut T, &AcceptContext<'_>, &ArgParser<'_>);
-type AcceptMapping<T> = &'static [(&'static [rustc_span::Symbol], AcceptFn<T>)];
+type AcceptMapping<T> = &'static [(Symbol, AcceptFn<T>)];
```
We can't do this. There are attributes that require more than one symbol in its path, and doing this would mean we can't parse those anymore, which would be a shame. None of the first few parsers I built use this feature, but I think it is an important one

jieyouxu replied:
Like namespaced attributes?

jdonszelmann replied:
precisely

jdonszelmann reviewed (commented):
This is not a full review, was just skimming through, but I don't think simplifying the parsers to take one symbol is that nice since attributes can be namespaced. We don't do this a whole lot right now, but I wish we did more. Diagnostics attributes use this for example.

jdonszelmann commented:
we briefly discussed this during rustweek I believe, but coming back to this: I'm happy to accept all the comment changes and some more, but none of the symbol slice -> single symbol changes.

jieyouxu commented:
@rustbot author (re. direction)

nnethercote committed:
Avoid `rustc_span::` qualifiers.

In several files they are entirely unnecessary, with the relevant names
already imported. And in a few I have added the necessary `use` item.

nnethercote commented:
@jdonszelmann: I have removed the commit that changed the symbol slice to a single symbol. I have added a new commit that removes some unnecessary/undesirable `rustc_span::` qualifiers.

@rustbot ready

bors commented:
:pushpin: Commit 354b1cbcca13dc85bd0a57f8c2b254831394b6fb has been approved by `jdonszelmann`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jdonszelmann 354b1cbcca13dc85bd0a57f8c2b254831394b6fb -->
<!-- homu: {"type":"Approved","sha":"354b1cbcca13dc85bd0a57f8c2b254831394b6fb","approver":"jdonszelmann","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jdonszelmann commented:
@bors rollup
